programs/token/src/state/token.rs

```diff
@@ -0,0 +1,128 @@
+use crate::ID;
+
+use super::AccountState;
+use pinocchio::{
+    account_info::{AccountInfo, Ref},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+pub struct TokenAccount(*const u8);
+
+impl TokenAccount {
+    pub const LEN: usize = 165;
+
+    /// Performs owner and length validation on `AccountInfo` and returns a `Ref<T>` for safe borrowing.
+    pub fn from_account_info(
```

Comment:
can't I do the following:

```
let token = TokenAccount::from_account_info(empty_account); // no unsafe
token.amount() // <- this is UB
token.authority() // <- this returns random garbage
if token.is_native() { // this is UB and could do anything
    ...
} else {
    ...
}
```

and the same with an account that doesn't respect the expected account data layout

Reply:
nevermind I'm extremely dumb you _are_ checking the len I need to sleep